### DISCUSSION TITLE
Realtime in dart SDK

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2663] [Q&A]
Realtime time not working in Dart SDK. Is there any example of it?

**Comment by ganigeorgiev** [comment]
Note that the realtime apis (including the OAuth2 all-in-one call) doesn't work with Flutter Web at the moment. For this and other limitations you can check https://github.com/pocketbase/dart-sdk#limitations. If you are using Flutter Web just as an emulator/test I recommend building for desktop.

In general you can find Dart SDK examples in the "API Preview" section for each collection in the Admin UI and/or in the [readme](https://github.com/pocketbase/dart-sdk#services) of the package.

**Reply by Oggy4th** [reply]
When realtime apis will available to use?

**Reply by ganigeorgiev** [reply]
They are already supported for mobile and desktop.

They don't work only when using Flutter Web. There are existing workarounds (eg. using `dart:html`) but it is better to wait the upstream `dart-lang/http` package to add support for streamed responses - https://github.com/dart-lang/http/issues/595.

**Reply by Oggy4th** [reply]
[pocketbase/dart-sdk#11](https://github.com/pocketbase/dart-sdk/issues/11#issuecomment-1580951658) can I use dart-lang/sse like mentioned here?

**Reply by ganigeorgiev** [reply]
Not as part of the SDK, meaning that you'll have to handle the [Realtime APIs](https://pocketbase.io/docs/api-realtime/) on your own.

But is Flutter Web really required for your use case? Could you elaborate how do you use it?

The last time I experimented with it more in depth (~6months ago) the performance was very poor and it is certainly not suitable for a generic web app.

**Reply by Oggy4th** [reply]
I'm like 70% done writing my admin panel in flutter web. I wanted to have same codebase like models and utilities in both admin panel web and client app and just realized realtime doesn't work

**Reply by Oggy4th** [reply]
I can use js sdk in the routes that I need realtime and flutter web in other pages, it might work with the new embedding feature or i'll just serve them on different routes

**Reply by ganigeorgiev** [reply]
@Oggy4th I've pushed a small fix in [Dart SDK v0.10.1](https://pub.dev/packages/pocketbase) that allows providing a custom `httpClientFactory` (both for the regular and SSE requests). You can use that option with a custom client like [`fetch_client`](https://github.com/Zekfad/fetch_client). For example:

```dart
import "package:pocketbase/pocketbase.dart";
import "package:fetch_client/fetch_client.dart";
import "package:flutter/foundation.dart" show kIsWeb;

void main() {
  final pb = PocketBase(
    "http://127.0.0.1:8090",
    // load the fetch_client only for web, otherwise - fallback to the default http.Client()
    // optionally, you can also create a separate PocketBase instance just for the realtime events
    httpClientFactory: kIsWeb ? () => FetchClient(mode: RequestMode.cors) : null,
  );

  // ...
}
```

_Note that I only briefly tested the realtime events with the FetchClient package and I cannot guarantee that everything else works correctly._

**Reply by Oggy4th** [reply]
@ganigeorgiev

Hello, there

I've found the problem with Flutter Web realtime.

Like you said I am using fetch_client.

First, Flutter Web realtime called multiple times at the same time. 

![Screenshot_20230626_005904_com.sec.android.app.sbrowser_edit_629258268180542.jpg](https://github.com/pocketbase/pocketbase/assets/44497354/62698763-2f7c-4b74-a408-c2a36c2b5d92)

Second, When I upload files by http.Multipartfile, gives `Failed to execute fetch`.

![Screenshot_20230626_011747.jpg](https://github.com/pocketbase/pocketbase/assets/44497354/15128571-1e31-4d37-b572-00b5a645f770)

I think both errors cause by the `fetch_client`.

Is there any way to fix it or May you check this problem?

**Reply by ganigeorgiev** [reply]
I'm not able to reproduce the above errors with PocketBase Dart SDK v0.10.2 and fetch_client v1.0.0.

The realtime connection is logged only once and uploading a file seems to work fine for me. I've tested with `webdev serve` directly in the browser. I'm not sure if Flutter Web does something differently under the hood but I expect it to behave the same:

```html
// /example/index.html

<!DOCTYPE html>
<html>
  <head>
    <script defer src="example.dart.js"></script>
  </head>
  <body>
    <button id="create_btn">Create record</button>
  </body>
</html>
```

```dart
// /example/example.dart
// dart pub global run webdev serve example

import 'dart:html';

import "package:pocketbase/pocketbase.dart";
import 'package:fetch_client/fetch_client.dart';
import 'package:http/http.dart' as http;

void main() {
  final pb = PocketBase(
    "http://127.0.0.1:8090",
    httpClientFactory: () => FetchClient(mode: RequestMode.cors),
  );

  pb.collection("demo").subscribe("*", (e) {
    print(e);
  });

  document.getElementById('create_btn')?.onClick.listen((event) {
    pb.collection("demo").create(files: [
      http.MultipartFile.fromString(
        'file',
        'example document...',
        filename: 'example_document.txt',
      ),
    ]).then((result) {
      print("Created $result");
    }).catchError((e) {
      print("ERROR: $e");
    });
  });
}
```

**Reply by shurentuya** [reply]
I'm having same issue, I think it only occurs when I'm trying to upload files larger than 1mb

**Reply by ganigeorgiev** [reply]
@shurentuya Is there any specific error thrown? Could you provide a code sample how do you initialize the client and send the request?

Note that if PocketBase is behind a reverse proxy (nginx, apache, caddy, etc.) the 1mb restriction could be from a default proxy configuration (eg. the default nginx [`client_max_body_size`](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size) is 1mb).

**Reply by shurentuya** [reply]
It is not behind proxy, error is `HttpException` from `http` package
```
ClientException: {url: https://url/api/collections/test/records/3gc4a6fn6hgi4nh, isAbort: true, statusCode: 0, response: {}, originalError: Failed to execute fetch: TypeError: Failed to fetch}
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 288:49  throw_
packages/pocketbase/src/client.dart 200:9                                     send
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 60:31            <fn>
dart-sdk/lib/async/zone.dart 1666:54                                          runBinary
dart-sdk/lib/async/future_impl.dart 162:22                                    handleError
dart-sdk/lib/async/future_impl.dart 796:46                                    handleError
dart-sdk/lib/async/future_impl.dart 817:13                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 592:5                                     [_completeError]
dart-sdk/lib/async/future_impl.dart 683:7                                     callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                              _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                               _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 177:15           <fn>
```
I think your minimal code will throw the same error if you increase the file size

**Reply by ganigeorgiev** [reply]
Indeed there seems to be some issue with the file size when using `fetch_client`. In my case with ~3mb encoded string the create request remains in "pending" state and doesn't seem to complete. With the default `dart-lang/http` it seems to work fine, I'm not sure what is causing it to freeze. I see that you opened an issue in the `fetch_client` repo (https://github.com/Zekfad/fetch_client/issues/6) so lets wait to see whether the author has some pointers.

As a temp workaround for Flutter Web and the realtime subscriptions, you can initialize 2 PocketBase instances - one with `fetch_client` to handle only the realtime subscriptions and another one with the default `dart-lang/http` for the regular http requests.

**Reply by ganigeorgiev** [reply]
@Oggy4th, @shurentuya As mentioned in https://github.com/Zekfad/fetch_client/issues/6#issuecomment-1615961604, the issue with the file upload and the `fetch_client` should be resolved in [Dart SDK v0.10.3](https://pub.dev/packages/pocketbase/versions/0.10.3).
